src/build_tools/hooks_hydra.py

```diff
@@ -141,6 +146,46 @@ def parse_hook(ec, *args, **kwargs):  # pylint: disable=unused-argument
         ec.log.info(f"[parse hook] Set parameter cuda_compute_capabilities: {ec['cuda_compute_capabilities']}")
 
 
+def pre_fetch_hook(self):
+    """Hook at pre-fetch level"""
+
+    # check/wait for fetch lock
+    source_path = source_paths()[0]
+    full_source_path = os.path.join(source_path, letter_dir_for(self.name), self.name)
+    lock_name = full_source_path.replace('/', '_') + '.lock'
+    lock_path = os.path.join(source_path, '.locks', lock_name)
+    self.fetch_hook_lock_path = lock_path
+
+    wait_interval = 60
+    wait_time = 0
+    wait_limit = 60 * 30
+
+    if os.path.exists(lock_path):
+        self.log.debug("[pre-fetch hook] Lock %s exists, waiting max %d seconds...", lock_path, wait_limit)
+        while os.path.exists(lock_path) and wait_time < wait_limit:
+            time.sleep(wait_interval)
+            wait_time += wait_interval
+
+        if os.path.exists(lock_path) and wait_time >= wait_limit:
+            error_msg = "[pre-fetch hook] Maximum wait time for lock %s to be released reached: %s sec >= %s sec"
+            raise EasyBuildError(error_msg, lock_path, wait_time, wait_limit)
+
+        self.log.info("[pre-fetch hook] Lock %s was released", lock_path)
+
+    # create fetch lock
+    mkdir(lock_path, parents=True)
```

Comment:
This is on a NFS export. I'm not sure how this will behave at scale...
Not sure if it's a good idea to use a directory/file as lock. Caching might be an issue here too.

Can't we (ab)use a singleton job for this?

Reply:
we can store the locks in scratch if that's more robust?
imo it's not a big issue to resubmit the build for a failed fetch as long as it doesn't happen too often.

using singleton jobs, you are either limited to building for one architecture at a time, or you have to split every job into a fetch job and a build job, which i wanted to avoid.

Reply:
i can also make the lockdir creation more atomic by using `os.mkdir(lock_path)` directly instead of first checking the existance and then using the EB function `mkdir` to create it.

Reply:
I prefer to keep it simple and use a directory/file lock. EasyBuild already uses that mechanism for software installations where the lock is placed inside the NFS mount in  `/apps/brussel/RL8/` and I have not seen any issues with it.

Moreover, if we want to upstream this to EB, it is better if we use the same mechanism.

Reply:
a fetch lock is however more likely to hit than the EB install lock, as all architecture build jobs are expected to fetch the same sources are around the same time.

i stress tested the `flufl.lock` package (https://pypi.org/project/flufl.lock/) which is designed for safe locking on NFS. seems to work nicely.
downside of using this package is that it's not suitable for upstream, but i'm ok with that considering how easy it was to implement in the hooks.

Reply:
i ran the same stress test using a simple dir lock, and that worked flawless too..

Reply:
if it is simple, that's fine for me :slightly_smiling_face:

Reply:
Locking is hard thing to do right in general. If there is an easy upstream library to do it, I would prefer we use that one.